### PR TITLE
Quark v2

### PR DESCRIPTION
**Pull request by kevincheng96**
This is a moderately-sized redesign of Quark v1. The main goal of this redesign is to improve the security of replayable operations by making them only submittable by submitters that have access to a secret submission token. A secondary goal of this redesign is to simplify a bunch of the flows to reduce gas overhead and code complexity. 

Changes:
- Introduce new replay mechanism that requires a submitter to provide a `submissionToken`. Previously, replayable operations were submittable by any address
- Remove isolated storage in the `QuarkStateManager`, which allows us to remove much of the complicated codepaths related to `setActiveNonceAndCallback`, `activeNonceScript`, `nonceScriptAddress`, and `walletStorage`
- Rename `QuarkStateManager` to `QuarkNonceManager` since its job is now to only manage nonces
- Adjust `QuarkOperation` struct to have a `isReplayable` field
- Bump up Solidity version to 0.8.27 and EVM version to `cancun`
- Use `TSTORE/TLOAD` to save gas

### CODE REVIEW

**Review by fluffywaffles — APPROVED** [review]
🙌

File: src/quark-core/src/QuarkNonceManager.sol [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        bytes32 lastTokenSubmission = submissions[msg.sender][nonce];
+        if (lastTokenSubmission == EXHAUSTED) {
+            revert NonReplayableNonce(msg.sender, nonce, submissionToken);
+        }
+        // Defense-in-depth check for `nonce != FREE` and `nonce != EXHAUSTED`
+        if (nonce == FREE || nonce == EXHAUSTED) {
+            revert InvalidNonce(msg.sender, nonce);
+        }
+        // Defense-in-depth check for `submissionToken != FREE` and `submissionToken != EXHAUSTED`
+        if (submissionToken == FREE || submissionToken == EXHAUSTED) {
```
Comment (hayesgm) [inline src/quark-core/src/QuarkNonceManager.sol:60]: It's fun/funny if you use `nonce=sha3(exhausted)`, you'll get an incidentally non-replayable operation. I think it's fine and probably quite good, just funny.

File: test/quark-core/EIP712.t.sol [hunk 2]
```diff
@@ -270,45 +277,46 @@ contract EIP712Test is Test {
...
         assertEq(counter.number(), 0);
-        assertEq(stateManager.nextNonce(address(wallet)), op.nonce);
+        assertEq(nonceManager.submissions(address(wallet), op.nonce), bytes32(uint256(0)));
     }
 
-    function testNonceIsNotSetForReplayableOperation() public {
-        // gas: do not meter set-up
-        vm.pauseGasMetering();
-        bytes memory incrementer = new YulHelper().getCode("Incrementer.sol/Incrementer.json");
+    // TODO: Uncomment when replay tokens are supported
```
Comment (hayesgm) [inline test/quark-core/EIP712.t.sol:283]: @kevincheng96

Comment (kevincheng96) [reply test/quark-core/EIP712.t.sol]: This test is no longer needed. Removed

File: src/quark-core/src/QuarkNonceManager.sol (hunk 1 above)
Comment (kevincheng96) [reply src/quark-core/src/QuarkNonceManager.sol:60]: Yeah, your hash chain ends immediately if you somehow end up with `0x00...00` or `0xFF...FF` as the submission token.

**Pull request merged**